Gd 189

Barmalej24 opened the pull request:
#GD-189 state: To Verify work 30м
Fixed Domain Exception

Barmalej24 committed:
#GD-185 state: To Verify work 60м
Making changes to the class name (plural)
Change to primary constructors in controllers
Add documentation to services
Added overloads (== & !=) for value objects

Barmalej24 committed:
#GD-189 state: To Verify work 30м
Fixed Domain Exception

oleggolen commented on EmailSenderMicroservice.Application/Services/Abstraction/Base/IService.cs:
```diff
@@ -6,9 +6,9 @@
-        where TEntity : class 
-        where TEntityAdd : class 
+    public interface IService<TEntity, TEntityAdd, TKey>
+        where TEntity : class
+        where TEntityAdd : class
```
Необходимо сделать маркерные интерфейсы для TEntity и TEntityAdd, не смотря на то, что это записи, но всё-равно лучше получать в сервисы только то, что надо

Barmalej24 replied:
добавлено

oleggolen commented on EmailSenderMicroservice.Application/Services/MessageService.cs:
```diff
@@ -3,46 +3,57 @@
     {
-        private readonly IMessageRepository _messageRepository;
-        private readonly IMapper _mapper;
         private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
```
Почему не сделать CansellationToken параметром методов, который передавать стандартный от контроллера, и сделать его параметром по умолчанию?

Barmalej24 replied:
проброшен дефолтовый от контроллера

oleggolen commented on EmailSenderMicroservice.Application/Services/SettingService.cs:
```diff
@@ -2,19 +2,23 @@
+    public class SettingService(ISettingRepository settingRepository) : ISettingService
     {
-        private readonly ISettingRepository _settingRepository;
         private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
```
Сделать необходимо сделать CancellationToken параметром по умаолчанию методов сервиса, чтобы его передавать из контроллера, либо использовать дефолтный

Barmalej24 replied:
проброшен дефолтовый от контроллера

oleggolen commented on EmailSenderMicroservice.DataAccess/Repositories/MessageRepository.cs:
```diff
@@ -44,7 +44,6 @@ public async Task<IEnumerable<Message>> GetAllAsync(CancellationToken cancellati
         public async Task<Guid> AddAsync(Message entity, CancellationToken cancellationToken)
         {
-
             await context.Messages.AddAsync(entity);
             await context.SaveChangesAsync();
```
Почему не передавать CansellationToken в параметры этих методов и не сделать проверку на то, что не было заявлено отмены до начала вызова этих методов?

Barmalej24 replied:
добавлен

oleggolen commented on EmailSenderMicroservice.DataAccess/Repositories/SettingRepository.cs:
```diff
@@ -21,7 +21,7 @@ public async Task<Guid> AddAsync(Setting entity, CancellationToken cancellationT
         {
             await context.Settings.AddAsync(entity);
             await context.SaveChangesAsync();
```
Некобходимо проверять не сработал ли CansellationToken до вызова этих методов и передавать его в параметры этих методов

Barmalej24 replied:
добавлен

oleggolen commented on EmailSenderMicroservice.Domain/Entities/Message.cs:
```diff
@@ -19,12 +18,12 @@ public class Message : IEntity<Guid>
         /// <summary>
         /// Название шаблона сообщений
         /// </summary>
         public string MessageType { get; }
```
Почему это строка, а не какой-то объект или перечисление?

Barmalej24 replied:
Неизвестно какие типы будут добавлены, а так всегда строковое описание

oleggolen commented on EmailSenderMicroservice.Domain/Entities/Setting.cs:
Пароль лучше сделать ValueObject'ом

Barmalej24 replied:
реализовано

oleggolen reviewed (dismissed):
Уже всё отлично, осталось улучшить работу с CansellationToken'ами, но пока это не супер срочно, и можем пропустить "пустые" запросы на сервер, также добавить ValueObject и доделать маркерные интерфейсы, чтобы защититься от всякой ерунды в параметрах в виде объектов произвольных классаов

m-a-ilnitsky commented on EmailSenderMicroservice.Application/Services/MessageService.cs:
```diff
@@ -3,46 +3,57 @@
+    /// </summary>
+    /// <param name="messageRepository">Репозиторий для работы с сообщениями.</param>
+    /// <param name="mapper">Маппер для преобразования сущностей в модели и наоборот.</param>
+    public class MessageService(IMessageRepository messageRepository, IMapper mapper) : IMessageService
```
Мне кажется эта логика когда аргумент конструктора используется как поле сильно затрудняет восприятие кода.
Когда видишь в коде имена _messageRepository и _mapper понимаешь что это поля.
А когда видишь messageRepository и mapper то непонятно откуда берутся эти переменные, которые не объявлены ни как переменные, ни как аргументы конкретного метода, код которого смотришь.
Мне кажется в таких случаях всё-таки лучше делать приватные поля.
А синтаксис примари конструктора использовать только в простых случаях, когда он не запутывает, а наоборот упрощает чтение кода

Barmalej24 replied:
использования primery конструктора пожелание Олега, которое было высказано в прошлых PR.

m-a-ilnitsky commented on EmailSenderMicroservice.Application/Services/SenderService.cs:
```diff
@@ -4,30 +4,39 @@
     {
-        private readonly ISettingService _settingService;
-        private readonly IMessageService _messageService;
         private const string fromName = "Email Notification Service";
```
Имя константы лучше начинать с заглавной буквы, чтобы в коде было понятно что это не переменная или аргумент

Barmalej24 replied:
исправлено

m-a-ilnitsky commented on EmailSenderMicroservice.DataAccess/Configurations/MessageConfiguration.cs:
```diff
@@ -15,7 +15,7 @@ public void Configure(EntityTypeBuilder<Message> builder)
             builder.Property(x => x.Email)
                 .HasConversion(
-                    v => v.Value, 
+                    v => v.Value,
                     v => new Email(v))
```
Сейчас в первой лямбде имя аргумента (v) неправильное, т.к. это не само value, а объект ValueObject.
Здесь аргументам лямбд лучше давать разные имена, т.к. у них разный смысл.
В первом случае это объект ValueObject Email, а во втором строка.
Поэтому можно назвать o и s, или vo и s.
Или email и str (чтобы код лямбд был понятнее можно использовать и полные имена, когда не удается придумать хорошие короткие).

В целом это некритично когда лямбды простые и невложенные друг в друга.
Поэтому иногда для быстроты написания в таких случаях используют одинаковые имена, но тогда эти имена стараются делать максимально абстрактными.
Т.е. здесь в обеих лямбдах вместо "v" (сокращение от value) можно было бы использовать "e" (сокращение от element)

Barmalej24 replied:
согласен.

Barmalej24 replied:
исправлено

m-a-ilnitsky commented on EmailSenderMicroservice.Domain/ValueObjects/Connection.cs:
```diff
@@ -106,12 +105,34 @@ public override bool Equals(object obj)
+        /// <returns><c>true</c>, если оба объекта равны; в противном случае <c>false</c>.</returns>
+        public static bool operator ==(Connection left, Connection right)
+        {
+            if (ReferenceEquals(left, null))
+                return ReferenceEquals(right, null);
+            return left.Equals(right);
```
Вместо: ReferenceEquals(left, null)
можно писать более кратко: left is null

И здесь можно не реализовывать проверки на null самостоятельно, а просто использовать статический Equals:
return Equals(left, right);

Barmalej24 replied:
ReferenceEquals(left, null) - проверка дает что объект точно не пустой, т.е не ссылается на соответствующую ячейку памяти

m-a-ilnitsky replied:
Это понятно, но left is null даёт тот же самый результат.
И если делать вызов статического метода (return Equals(left, right);) то не важно равно ли одно или даже оба значения null, статический Equals из-за этого не упадет и в любом случае выдаст результат

m-a-ilnitsky commented on EmailSenderMicroservice.Domain/ValueObjects/Email.cs:
```diff
@@ -0,0 +1,94 @@
+        /// Возвращает хеш-код для текущего объекта <see cref="Email"/>.
+        /// </summary>
+        /// <returns>Хеш-код для текущего объекта.</returns>
+        public override int GetHashCode() => StringComparer.Ordinal.GetHashCode(Value);
```
Скорее всего вместо этого: StringComparer.Ordinal.GetHashCode(Value);
можно делать так: Value.GetHashCode(StringComparison.Ordinal);
Но тогда нужно это проверять, что будет тоже самое.
Никогда такого не делал.
В целом, мне не понятно почему не использовать обычный Value.GetHashCode();

Barmalej24 replied:
насколько я понял StringComparer.Ordinal.GetHashCode(Value) генерирует хеш на основе последовательности символов, не беря во внимание возможные изменения значения или время их создания в процессе выполнения.

m-a-ilnitsky replied:
Блинн, а какое отношение время создания имеет к значению строки...
Как я понимаю GetHashCode и так генерит значение только на основании последовательности символов строки.
Эти навороты скорее всего нужны только если хочется сравнивать строки без учета регистра символов и тогда нужна соответствующая логика при вычислении хэша

m-a-ilnitsky reviewed (dismissed):
В целом код хороший.
Но лично мне использование нового синтаксиса когда вместо приватных полей используются аргументы первичного конструктора показалось плохой практикой, когда класс длиннее половины экрана, т.к. когда в коде какого-то метода используется _messageRepository то понятно что это поле объявленное в классе, а когда вместо него используется messageRepository, то уже непонятно где искать объявление этой переменной/аргумента.
В коротких классах это не проблема, а вот в длинных я бы всё-таки делал приватные поля

Barmalej24 committed:
Added ValueObj Password
Added missing cancellation token in repository methods
Corrected symbols in lambdas
Local cancellation token has been removed in Services and forwarded from the controller
Added marker interfaces for models
